services/code_hosting/git_hub_service.rb

```diff
@@ -1,131 +1,421 @@
+      # to fast things, as each client logs in using netrc.
+      # https://github.com/octokit/octokit.rb/blob/a06d16359ca3b529aea42ca4d84f9a4fc99de0dd/lib/octokit/client.rb#L123
+      # @param [String] api_token
+      def client(api_token)
+        @client_cache ||= {}
+        return @client_cache[api_token] unless @client_cache[api_token].nil?
+        client = Octokit::Client.new(access_token: api_token)
+        @client_cache[api_token]
+        return client
+      end
+
+      def temp_path(root_path = Dir.tmpdir)
+        temporary_path = File.join(root_path, ".fastlane")
+        FileUtils.mkdir_p(temporary_path) unless File.directory?(temporary_path)
+        return temporary_path
+      end
+
+      def temporary_git_storage(root_path = Dir.tmpdir)
+        temp_storage = File.join(root_path, ".tmp")
+        @temporary_git_storage ||= temp_storage
+        FileUtils.mkdir_p(@temporary_git_storage) unless File.directory?(@temporary_git_storage)
+        return @temporary_git_storage
+      end
+
+      # This method shallow clones the repo using the provider credential and looks for a valid _fastlane_
+      # configuration, parses the Fastfile (if found), and returns the lanes in the following way:
+      #  {
+      #    "no_platform": {
+      #      "": {
+      #        "description": [
+      #
+      #        ],
+      #        "actions": [
+      #          {
+      #            "action": "default_platform",
+      #            "parameters": "ios"
+      #          }
+      #        ]
+      #      }
+      #    },
+      #    "ios": {
+      #      "test": {
+      #        "description": [
+      #          "Description of what the lane does"
+      #        ],
+      #        "actions": [
+      #          {
+      #            "action": "gym",
+      #            "parameters": {
+      #              "skip_package_ipa": false,
+      #              "clean": true,
+      #              "project": "../fastlane-ci-demoapp.xcodeproj",
+      #              "scheme": "fastlane-ci-demoapp"
+      #            }
+      #          }
+      #        ],
+      #        "private": false
+      #      }
+      #    }
+      #  }
+      #
+      # @param repo_url [String]
+      # @param branch [String]
+      # @param path [String]
+      # @param provider_credential [GithubProviderCredential]
+      # @return [(Fastlane::FastfileParser, Hash<String, Hash>)] the FastlaneParser instance used for the parsing and a hash, being the key the name of the platform (:ios, :android, :no_platform) and the Hash the underlying lane (name, actions).
+      def peek_fastfile_configuration(repo_url: nil, branch: nil, provider_credential: nil, path: self.temp_path, cache: true)
+        repo = repo_from_url(repo_url)
+        # If cache is allowed for this call, and there's a Hash under the key, use the cache.
+        if cache && self.cache&.[]([repo, branch].join("/"))&.kind_of?(Hash)
```

Comment:
I think this is safer, shorter and prettier 🎨

Reply:
It's pretty, butttt I have no idea what it's doing 🙃 
I'm sure you know this is coming:
I know it's clever, but it's really confusing and would be easier for new people (and me) to understand if you broke it into multiple lines with intermediate temporary variables describing the steps.

Reply:
Gonna split it! 😄 But just in case you're not familiar with the `&.` is the safe operator chaining in Ruby (I don't know if you're familiar with Swift, but it's like the optional chaining `?.`) So long history short, calls the chain of methods until a `nil` is received. I don't know if this is a feature that I should be using systematically (I think this adds a ton of code safety).
After the explanation, do I split this? I have no problem, I adapt to your point of view 😄